File: src/services/gmail.ts
/**
 * Gmail Integration Service
 * 
 * Ported from quinn-v2/services/gmail.py with modifications:
 * - Remove QUINN_EMAILS constant 
 * - Remove INBOX_MAP
 * - Generalize is_from_quinn → is_from_sender
 * - TypeScript implementation
 * - GMAIL_ENABLED=false by default (stub mode)
 */

import { PrismaClient, SignalType } from '@prisma/client';
import { google } from 'googleapis';

const prisma = new PrismaClient();

export interface GmailMessage {
  id: string;
  threadId: string;
  snippet: string;
  payload: any;
  internalDate: string;
}

export interface ParsedEmail {
  messageId: string;
  threadId: string;
  subject: string;
  from: string;
  to: string;
  date: Date;
  body: string;
  isReply: boolean;
  replyToMessageId?: string;
  signalType?: SignalType;
}

export class GmailService {
  private oauth2Client: any;
  private gmail: any;

  constructor() {
    if (process.env.GMAIL_ENABLED === 'true') {
      // Initialize Google OAuth2 client
      this.oauth2Client = new google.auth.OAuth2(
        process.env.GOOGLE_CLIENT_ID,
        process.env.GOOGLE_CLIENT_SECRET,
        process.env.GOOGLE_REDIRECT_URI
      );

      this.gmail = google.gmail({ version: 'v1', auth: this.oauth2Client });
    }
  }

  /**
   * Set access credentials for a mailbox
   */
  async setCredentials(mailboxId: string): Promise<boolean> {
    if (process.env.GMAIL_ENABLED !== 'true') {
      console.log(`[STUB] Setting credentials for mailbox: ${mailboxId}`);
      return true;
    }

    try {
      const mailbox = await prisma.mailbox.findUnique({
        where: { id: mailboxId },
        select: {
          accessToken: true,
          refreshToken: true,
          tokenExpiresAt: true,
        },
      });

      if (!mailbox || !mailbox.refreshToken) {
        throw new Error('Mailbox not found or missing refresh token');
      }

      this.oauth2Client.setCredentials({
        access_token: mailbox.accessToken,
        refresh_token: mailbox.refreshToken,
        expiry_date: mailbox.tokenExpiresAt?.getTime(),
      });

      // Refresh token if expired
      if (!mailbox.accessToken || (mailbox.tokenExpiresAt && mailbox.tokenExpiresAt < new Date())) {
        const { credentials } = await this.oauth2Client.refreshAccessToken();
        
        // Update stored tokens
        await prisma.mailbox.update({
          where: { id: mailboxId },
          data: {
            accessToken: credentials.access_token,
            tokenExpiresAt: new Date(credentials.expiry_date),
          },
        });

        this.oauth2Client.setCredentials(credentials);
      }

      return true;
    } catch (error) {
      console.error(`Failed to set Gmail credentials for mailbox ${mailboxId}:`, error);
      return false;
    }
  }

  /**
   * Send an email via Gmail API
   */
  async sendEmail(params: {
    mailboxId: string;
    to: string;
    toName?: string;
    subject: string;
    body: string;
    replyToMessageId?: string;
  }): Promise<{ messageId: string; success: boolean }> {
    if (process.env.GMAIL_ENABLED !== 'true') {
      // Stub mode
      const messageId = `stub-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
      console.log(`[STUB] Sending email via Gmail: ${params.subject} to ${params.to}`);
      return { messageId, success: true };
    }

    try {
      const credentialsSet = await this.setCredentials(params.mailboxId);
      if (!credentialsSet) {
        throw new Error('Failed to set Gmail credentials');
      }

      const mailbox = await prisma.mailbox.findUnique({
        where: { id: params.mailboxId },
        select: { email: true, displayName: true },
      });

      if (!mailbox) {
        throw new Error('Mailbox not found');
      }

      // Construct email message
      const messageParts = [
        `To: ${params.toName ? `"${params.toName}" <${params.to}>` : params.to}`,
        `From: ${mailbox.displayName ? `"${mailbox.displayName}" <${mailbox.email}>` : mailbox.email}`,
        `Subject: ${params.subject}`,
        'Content-Type: text/html; charset=UTF-8',
        'MIME-Version: 1.0',
        '',
        params.body
      ];

      if (params.replyToMessageId) {
        messageParts.splice(3, 0, `In-Reply-To: <${params.replyToMessageId}>`);
      }

      const message = messageParts.join('\n');
      const encodedMessage = Buffer.from(message).toString('base64')
        .replace(/\+/g, '-')
        .replace(/\//g, '_')
        .replace(/=+$/, '');

      const response = await this.gmail.users.messages.send({
        userId: 'me',
        requestBody: {
          raw: encodedMessage,
        },
      });

      return { messageId: response.data.id, success: true };
    } catch (error) {
      console.error(`Failed to send email via Gmail:`, error);
      return { messageId: '', success: false };
    }
  }

  /**
   * Check for new messages and signals
   */
  async checkForSignals(mailboxId: string, lastCheckTime?: Date): Promise<ParsedEmail[]> {
    if (process.env.GMAIL_ENABLED !== 'true' && process.env.SCOUT_STUB_MODE !== 'true') {
      return [];
    }

    if (process.env.SCOUT_STUB_MODE === 'true') {
      // Return stub signals for testing
      return this.generateStubSignals(mailboxId);
    }

    try {
      const credentialsSet = await this.setCredentials(mailboxId);
      if (!credentialsSet) {
        return [];
      }

      // Build Gmail API query
      let query = 'in:inbox';
      if (lastCheckTime) {
        const timestamp = Math.floor(lastCheckTime.getTime() / 1000);
        query += ` after:${timestamp}`;
      }

      const response = await this.gmail.users.messages.list({
        userId: 'me',
        q: query,
        maxResults: 50,
      });

      if (!response.data.messages) {
        return [];
      }

      const parsedEmails: ParsedEmail[] = [];

      for (const message of response.data.messages) {
        try {
          const fullMessage = await this.gmail.users.messages.get({
            userId: 'me',
            id: message.id,
          });

          const parsed = await this.parseGmailMessage(fullMessage.data, mailboxId);
          if (parsed) {
            parsedEmails.push(parsed);
          }
        } catch (error) {
          console.error(`Failed to parse message ${message.id}:`, error);
        }
      }

      return parsedEmails;
    } catch (error) {
      console.error(`Failed to check for Gmail signals:`, error);
      return [];
    }
  }

  /**
   * Parse a Gmail message and determine signal type
   */
  private async parseGmailMessage(message: GmailMessage, mailboxId: string): Promise<ParsedEmail | null> {
    try {
      const headers = message.payload.headers || [];
      const getHeader = (name: string) => headers.find((h: any) => h.name.toLowerCase() === name.toLowerCase())?.value || '';

      const subject = getHeader('subject');
      const from = getHeader('from');
      const to = getHeader('to');
      const date = new Date(parseInt(message.internalDate));
      const messageId = getHeader('message-id');
      const inReplyTo = getHeader('in-reply-to');
      const references = getHeader('references');

      // Extract email body
      const body = this.extractEmailBody(message.payload);

      // Determine if this is a reply to one of our sent emails
      const isReply = await this.isReplyToSentEmail(mailboxId, inReplyTo, references, from);
      
      // Detect signal type
      let signalType: SignalType | undefined;
      
      if (isReply) {
        if (this.isOutOfOfficeReply(subject, body)) {
          signalType = 'OUT_OF_OFFICE';
        } else if (this.isBounceMessage(from, subject, body)) {
          signalType = 'BOUNCE';
        } else if (this.isUnsubscribeMessage(subject, body)) {
          signalType = 'UNSUBSCRIBE';
        } else {
          signalType = 'REPLY';
        }
      }

      return {
        messageId: message.id,
        threadId: message.threadId,
        subject,
        from,
        to,
        date,
        body,
        isReply,
        replyToMessageId: inReplyTo,
        signalType,
      };
    } catch (error) {
      console.error('Failed to parse Gmail message:', error);
      return null;
    }
  }

  /**
   * Extract plain text body from Gmail message payload
   */
  private extractEmailBody(payload: any): string {
    if (payload.body?.data) {
      return Buffer.from(payload.body.data, 'base64').toString('utf-8');
    }

    if (payload.parts) {
      for (const part of payload.parts) {
        if (part.mimeType === 'text/plain' && part.body?.data) {
          return Buffer.from(part.body.data, 'base64').toString('utf-8');
        }
        
        // Recursively check nested parts
        if (part.parts) {
          const nestedBody = this.extractEmailBody(part);
          if (nestedBody) return nestedBody;
        }
      }
    }

    return '';
  }

  /**
   * Check if message is a reply to one of our sent emails
   * Generalized from is_from_quinn to is_from_sender
   */
  private async isReplyToSentEmail(mailboxId: string, inReplyTo: string, references: string, fromEmail: string): Promise<boolean> {
    if (!inReplyTo && !references) {
      return false;
    }

    // Check if we have a sent email with matching message ID
    const messageIds = [inReplyTo, ...(references ? references.split(/\s+/) : [])].filter(Boolean);
    
    if (messageIds.length === 0) {
      return false;
    }

    const sentEmail = await prisma.sentEmail.findFirst({
      where: {
        mailboxId,
        messageId: {
          in: messageIds,
        },
      },
    });

    return !!sentEmail;
  }

  /**
   * Check if message is an out-of-office reply
   */
  private isOutOfOfficeReply(subject: string, body: string): boolean {
    const oofPatterns = [
      /out of office/i,
      /away from office/i,
      /vacation/i,
      /holiday/i,
      /absence/i,
      /unavailable/i,
      /automatic reply/i,
      /auto-reply/i,
    ];

    const text = `${subject} ${body}`;
    return oofPatterns.some(pattern => pattern.test(text));
  }

  /**
   * Check if message is a bounce/delivery failure
   */
  private isBounceMessage(from: string, subject: string, body: string): boolean {
    const bounceFromPatterns = [
      /mailer-daemon/i,
      /postmaster/i,
      /noreply/i,
      /no-reply/i,
    ];

    const bounceSubjectPatterns = [
      /delivery.*fail/i,
      /undeliverable/i,
      /bounce/i,
      /returned mail/i,
      /delivery status notification/i,
    ];

    const bounceBodyPatterns = [
      /delivery.*fail/i,
      /recipient.*unknown/i,
      /user.*not.*found/i,
      /mailbox.*full/i,
      /message.*rejected/i,
    ];

    const text = `${subject} ${body}`;
    
    return bounceFromPatterns.some(pattern => pattern.test(from)) ||
           bounceSubjectPatterns.some(pattern => pattern.test(subject)) ||
           bounceBodyPatterns.some(pattern => pattern.test(text));
  }

  /**
   * Check if message is an unsubscribe request
   */
  private isUnsubscribeMessage(subject: string, body: string): boolean {
    const unsubPatterns = [
      /unsubscribe/i,
      /remove.*list/i,
      /stop.*email/i,
      /opt.*out/i,
    ];

    const text = `${subject} ${body}`;
    return unsubPatterns.some(pattern => pattern.test(text));
  }

  /**
   * Generate stub signals for testing
   */
  private generateStubSignals(mailboxId: string): ParsedEmail[] {
    const signals: ParsedEmail[] = [];
    
    // Randomly generate signals for demo
    if (Math.random() < 0.3) { // 30% chance
      const signalTypes: SignalType[] = ['REPLY', 'BOUNCE', 'OUT_OF_OFFICE', 'UNSUBSCRIBE'];
      const randomType = signalTypes[Math.floor(Math.random() * signalTypes.length)];
      
      signals.push({
        messageId: `stub-signal-${Date.now()}`,
        threadId: `thread-${Math.random().toString(36).substr(2, 9)}`,
        subject: this.getStubSubjectForSignalType(randomType),
        from: 'test@example.com',
        to: 'sender@example.com',
        date: new Date(),
        body: this.getStubBodyForSignalType(randomType),
        isReply: true,
        signalType: randomType,
      });
    }

    return signals;
  }

  private getStubSubjectForSignalType(signalType: SignalType): string {
    switch (signalType) {
      case 'OUT_OF_OFFICE':
        return 'Out of Office Auto-Reply';
      case 'BOUNCE':
        return 'Delivery Status Notification (Failure)';
      case 'UNSUBSCRIBE':
        return 'Please remove me from your list';
      case 'REPLY':
        return 'Re: Your email';
      default:
        return 'Test Signal';
    }
  }

  private getStubBodyForSignalType(signalType: SignalType): string {
    switch (signalType) {
      case 'OUT_OF_OFFICE':
        return 'I am currently out of the office and will return on Monday.';
      case 'BOUNCE':
        return 'The following message could not be delivered to the recipient.';
      case 'UNSUBSCRIBE':
        return 'Please unsubscribe me from this mailing list.';
      case 'REPLY':
        return 'Thanks for your email. I will get back to you soon.';
      default:
        return 'Test signal message body.';
    }
  }
}

export const gmailService = new GmailService();

File: src/services/gmail.py
"""
Gmail Service for Sequence Service.

Adapted from Quinn V2's gmail.py with modifications:
- Removed hardcoded Quinn emails
- Generalized for any mailbox via domain-wide delegation
- Simplified for sequence sending use case

Auth: Service account with domain-wide delegation
"""
import base64
import html
import re
import uuid
from datetime import datetime, timezone
from email.message import EmailMessage
from typing import Optional

from google.oauth2 import service_account
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError

from src.config import get_settings

settings = get_settings()


class GmailError(Exception):
    """Base exception for Gmail service errors."""
    pass


class GmailAuthError(GmailError):
    """Authentication or authorization error."""
    pass


class GmailAPIError(GmailError):
    """Gmail API error with status code and message."""
    
    def __init__(self, message: str, status_code: Optional[int] = None, reason: Optional[str] = None):
        self.message = message
        self.status_code = status_code
        self.reason = reason
        super().__init__(f"GmailAPIError [{status_code}]: {message}" if status_code else message)


class GmailService:
    """
    Gmail API integration service with domain-wide delegation.
    
    Can send from any mailbox that has been delegated to the service account.
    """
    
    SCOPES = [
        'https://www.googleapis.com/auth/gmail.send',
        'https://www.googleapis.com/auth/gmail.readonly',
    ]
    
    _instances: dict[str, 'GmailService'] = {}
    
    def __init__(self, inbox: str):
        """
        Initialize Gmail service for a specific inbox.
        
        Args:
            inbox: Email address to send from (must be delegated to service account)
        """
        self.inbox = inbox
        self._service = None
        self._credentials = None
    
    @property
    def service(self):
        """Lazy initialization of Gmail API service."""
        if self._service is None:
            self._service = self._build_service()
        return self._service
    
    def _build_service(self):
        """Build Gmail API service with delegated credentials."""
        if not settings.gmail_service_account_file:
            raise GmailAuthError(
                "gmail_service_account_file not configured. "
                "Set path to service account JSON file."
            )
        
        try:
            credentials = service_account.Credentials.from_service_account_file(
                settings.gmail_service_account_file,
                scopes=self.SCOPES
            )
            delegated_credentials = credentials.with_subject(self.inbox)
            self._credentials = delegated_credentials
            
            return build('gmail', 'v1', credentials=delegated_credentials)
        except FileNotFoundError:
            raise GmailAuthError(
                f"Service account file not found: {settings.gmail_service_account_file}"
            )
        except Exception as e:
            raise GmailAuthError(f"Failed to initialize Gmail service: {e}")
    
    @classmethod
    def get_inbox(cls, email: str) -> 'GmailService':
        """Get or create a GmailService instance for an inbox."""
        if email not in cls._instances:
            cls._instances[email] = cls(inbox=email)
        return cls._instances[email]
    
    def send_email(
        self,
        to: str,
        subject: str,
        body: str,
        thread_id: Optional[str] = None,
        message_id: Optional[str] = None,
        cc: Optional[str] = None,
        reply_to: Optional[str] = None,
        sender_name: Optional[str] = None,
    ) -> dict:
        """
        Send a plain text email.
        
        Args:
            to: Recipient email address
            subject: Email subject
            body: Plain text email body
            thread_id: Gmail thread ID for threading (reply)
            message_id: Message-ID header to reply to
            cc: CC recipients (comma-separated)
            reply_to: Reply-To header
            sender_name: Display name for sender
            
        Returns:
            dict with message_id, thread_id, label_ids
        """
        message = self._build_email_message(
            to=to,
            subject=subject,
            body=body,
            is_html=False,
            cc=cc,
            reply_to=reply_to,
            message_id=message_id,
            sender_name=sender_name,
        )
        
        return self._send_message(message, thread_id)
    
    def send_html_email(
        self,
        to: str,
        subject: str,
        html_body: str,
        thread_id: Optional[str] = None,
        message_id: Optional[str] = None,
        cc: Optional[str] = None,
        reply_to: Optional[str] = None,
        sender_name: Optional[str] = None,
        plain_text_fallback: Optional[str] = None,
        list_unsubscribe: Optional[str] = None,
        one_click: bool = False,
    ) -> dict:
        """Send an HTML email with optional plain text fallback."""
        message = self._build_email_message(
            to=to,
            subject=subject,
            body=html_body,
            is_html=True,
            cc=cc,
            reply_to=reply_to,
            message_id=message_id,
            sender_name=sender_name,
            plain_text_fallback=plain_text_fallback,
            list_unsubscribe=list_unsubscribe,
            one_click=one_click,
        )

        return self._send_message(message, thread_id)
    
    def _build_email_message(
        self,
        to: str,
        subject: str,
        body: str,
        is_html: bool = False,
        cc: Optional[str] = None,
        reply_to: Optional[str] = None,
        message_id: Optional[str] = None,
        sender_name: Optional[str] = None,
        plain_text_fallback: Optional[str] = None,
        list_unsubscribe: Optional[str] = None,
        one_click: bool = False,
    ) -> EmailMessage:
        """Build an EmailMessage object."""
        message = EmailMessage()

        if sender_name:
            message['From'] = f"{sender_name} <{self.inbox}>"
        else:
            message['From'] = self.inbox

        message['To'] = to
        message['Subject'] = subject

        if cc:
            message['Cc'] = cc
        if reply_to:
            message['Reply-To'] = reply_to
        if message_id:
            message['In-Reply-To'] = message_id
            message['References'] = message_id

        # RFC 8058 List-Unsubscribe headers. Only advertise One-Click when a
        # reachable HTTPS unsubscribe endpoint exists (one_click=True); otherwise
        # emit List-Unsubscribe (mailto) WITHOUT the -Post header, so we never
        # claim a one-click endpoint that fails (track.telnyx.com is NXDOMAIN).
        if list_unsubscribe:
            message['List-Unsubscribe'] = list_unsubscribe
            if one_click:
                message['List-Unsubscribe-Post'] = 'List-Unsubscribe=One-Click'

        if is_html:
            unique_token = f'<span style="color: #e0e0e0; font-size: 1px;">{uuid.uuid4()}</span>'
            html_with_token = f"{body}\n{unique_token}"
            
            if plain_text_fallback:
                message.set_content(plain_text_fallback)
            else:
                message.set_content(self._html_to_plain_text(body))
            
            message.add_alternative(html_with_token, subtype='html')
        else:
            message.set_content(body + '\u200B')
        
        return message
    
    def _send_message(self, message: EmailMessage, thread_id: Optional[str] = None) -> dict:
        """Send the email message via Gmail API."""
        try:
            raw_message = base64.urlsafe_b64encode(message.as_bytes()).decode('utf-8')
            body = {'raw': raw_message}
            
            if thread_id:
                body['threadId'] = thread_id
            
            result = self.service.users().messages().send(
                userId='me',
                body=body
            ).execute()
            
            return {
                'message_id': result.get('id'),
                'thread_id': result.get('threadId'),
                'label_ids': result.get('labelIds', []),
            }
            
        except HttpError as e:
            raise GmailAPIError(
                message=f"Failed to send email: {e.reason}",
                status_code=e.resp.status,
                reason=e.reason
            )
    
    @staticmethod
    def _html_to_plain_text(html_content: str) -> str:
        """Convert HTML to plain text (basic conversion)."""
        try:
            from bs4 import BeautifulSoup
            soup = BeautifulSoup(html_content, 'html.parser')
            return soup.get_text(separator='\n', strip=True)
        except ImportError:
            # Fallback without BeautifulSoup
            text = re.sub(r'<[^>]+>', '', html_content)
            return html.unescape(text)
    
    def check_connection(self) -> bool:
        """Test the Gmail API connection."""
        try:
            profile = self.service.users().getProfile(userId='me').execute()
            return profile.get('emailAddress') == self.inbox
        except HttpError as e:
            raise GmailAPIError(
                message=f"Connection check failed: {e.reason}",
                status_code=e.resp.status,
                reason=e.reason
            )
    
    def list_messages(
        self,
        query: str = "",
        max_results: int = 100,
        label_ids: Optional[list[str]] = None,
    ) -> list[dict]:
        """
        List messages matching a query.
        
        Args:
            query: Gmail search query (e.g., "is:unread", "from:user@example.com")
            max_results: Maximum messages to return
            label_ids: Filter by label IDs (e.g., ["INBOX"])
            
        Returns:
            List of message stubs with id and threadId
        """
        try:
            params = {
                'userId': 'me',
                'maxResults': max_results,
            }
            if query:
                params['q'] = query
            if label_ids:
                params['labelIds'] = label_ids
            
            result = self.service.users().messages().list(**params).execute()
            return result.get('messages', [])
            
        except HttpError as e:
            raise GmailAPIError(
                message=f"Failed to list messages: {e.reason}",
                status_code=e.resp.status,
                reason=e.reason
            )
    
    def get_message(self, message_id: str, format: str = "metadata") -> dict:
        """
        Get a single message.
        
        Args:
            message_id: Gmail message ID
            format: "minimal", "metadata", "full", or "raw"
            
        Returns:
            Message data
        """
        try:
            result = self.service.users().messages().get(
                userId='me',
                id=message_id,
                format=format,
            ).execute()
            return result
            
        except HttpError as e:
            raise GmailAPIError(
                message=f"Failed to get message: {e.reason}",
                status_code=e.resp.status,
                reason=e.reason
            )
    
    def get_thread(self, thread_id: str, format: str = "metadata") -> dict:
        """
        Get all messages in a thread.
        
        Args:
            thread_id: Gmail thread ID
            format: "minimal", "metadata", "full"
            
        Returns:
            Thread data with messages
        """
        try:
            result = self.service.users().threads().get(
                userId='me',
                id=thread_id,
                format=format,
            ).execute()
            return result
            
        except HttpError as e:
            raise GmailAPIError(
                message=f"Failed to get thread: {e.reason}",
                status_code=e.resp.status,
                reason=e.reason
            )
    
    def get_replies_to_threads(self, thread_ids: list[str]) -> list[dict]:
        """
        Get replies to specific threads (messages we sent).
        
        Returns messages that are NOT from our inbox address (i.e., replies from others).
        """
        replies = []
        
        for thread_id in thread_ids:
            try:
                thread = self.get_thread(thread_id, format="metadata")
                messages = thread.get('messages', [])
                
                for msg in messages:
                    # Get the From header
                    headers = {h['name'].lower(): h['value'] for h in msg.get('payload', {}).get('headers', [])}
                    from_addr = headers.get('from', '')
                    
                    # Skip messages from ourselves
                    if self.inbox.lower() in from_addr.lower():
                        continue
                    
                    # This is a reply from someone else
                    # Classify bounce/OOO here (F6): signal_detection reads
                    # reply['is_bounce']/['is_ooo']; without these keys every
                    # mailer-daemon bounce was recorded as a plain REPLY.
                    replies.append({
                        'message_id': msg['id'],
                        'thread_id': msg['threadId'],
                        'from': from_addr,
                        'subject': headers.get('subject', ''),
                        'date': headers.get('date', ''),
                        'snippet': msg.get('snippet', ''),
                        'label_ids': msg.get('labelIds', []),
                        'is_bounce': self.detect_bounce(msg),
                        'is_ooo': self.detect_out_of_office(msg),
                    })
                    
            except GmailAPIError:
                # Skip threads we can't access
                continue
        
        return replies
    
    def detect_bounce(self, message: dict) -> bool:
        """Check if a message is a bounce notification."""
        headers = {h['name'].lower(): h['value'] for h in message.get('payload', {}).get('headers', [])}
        from_addr = headers.get('from', '').lower()
        subject = headers.get('subject', '').lower()
        
        # Common bounce indicators
        bounce_senders = ['mailer-daemon', 'postmaster', 'mail-daemon']
        bounce_subjects = ['undeliverable', 'delivery status', 'delivery failed', 
                          'returned mail', 'failure notice', 'mail delivery failed']
        
        if any(sender in from_addr for sender in bounce_senders):
            return True
        if any(term in subject for term in bounce_subjects):
            return True
        
        return False
    
    def detect_out_of_office(self, message: dict) -> bool:
        """Check if a message is an out-of-office reply."""
        headers = {h['name'].lower(): h['value'] for h in message.get('payload', {}).get('headers', [])}
        subject = headers.get('subject', '').lower()
        snippet = message.get('snippet', '').lower()
        
        # Common OOO indicators
        ooo_terms = ['out of office', 'out of the office', 'automatic reply', 
                     'auto-reply', 'autoreply', 'away from', 'on vacation',
                     'currently out', 'limited access', 'ooo']
        
        text_to_check = subject + ' ' + snippet
        return any(term in text_to_check for term in ooo_terms)
    
    def get_new_inbox_messages(self, since_timestamp: Optional[datetime] = None) -> list[dict]:
        """
        Get new messages in inbox, optionally since a timestamp.
        
        Args:
            since_timestamp: Only get messages after this time
            
        Returns:
            List of message data with headers
        """
        query = "in:inbox"
        if since_timestamp:
            # Gmail uses epoch seconds for after: query
            epoch = int(since_timestamp.timestamp())
            query += f" after:{epoch}"
        
        message_stubs = self.list_messages(query=query, max_results=50)
        
        messages = []
        for stub in message_stubs:
            try:
                msg = self.get_message(stub['id'], format="metadata")
                headers = {h['name'].lower(): h['value'] for h in msg.get('payload', {}).get('headers', [])}
                
                messages.append({
                    'message_id': msg['id'],
                    'thread_id': msg['threadId'],
                    'from': headers.get('from', ''),
                    'to': headers.get('to', ''),
                    'subject': headers.get('subject', ''),
                    'date': headers.get('date', ''),
                    'in_reply_to': headers.get('in-reply-to', ''),
                    'references': headers.get('references', ''),
                    'snippet': msg.get('snippet', ''),
                    'label_ids': msg.get('labelIds', []),
                    'is_bounce': self.detect_bounce(msg),
                    'is_ooo': self.detect_out_of_office(msg),
                })
            except GmailAPIError:
                continue
        
        return messages


async def send_test_email(
    to: str,
    from_email: str,
    subject: str = "Test from Sequence Service",
    body: str = "This is a test email from the Sequence Service.",
    sender_name: Optional[str] = None,
) -> dict:
    """
    Send a test email.
    
    Args:
        to: Recipient email
        from_email: Sender email (must be delegated)
        subject: Email subject
        body: Email body
        sender_name: Display name
        
    Returns:
        Send result dict
    """
    gmail = GmailService.get_inbox(from_email)
    return gmail.send_email(
        to=to,
        subject=subject,
        body=body,
        sender_name=sender_name,
    )
